src/Marten/Internal/Sessions/EventTracingConnectionLifetime.cs

```diff
@@ -0,0 +1,286 @@
+    private const string MartenBatchPagesExecutionStarted = "MartenBatchPagesExecutionStarted";
+    private const string MartenBatchPagesExecutionFailed = "MartenBatchPagesExecutionFailed";
+
+    private readonly IConnectionLifetime _innerConnectionLifetime;
+    private readonly OpenTelemetryOptions _openTelemetryOptions;
+    private readonly Activity _databaseActivity;
+
+    public EventTracingConnectionLifetime(IConnectionLifetime innerConnectionLifetime, OpenTelemetryOptions openTelemetryOptions, Activity databaseActivity)
+    {
+        if (innerConnectionLifetime == null)
+        {
+            throw new ArgumentNullException(nameof(innerConnectionLifetime));
+        }
+
+        if (openTelemetryOptions == null)
+        {
+            throw new ArgumentNullException(nameof(openTelemetryOptions));
+        }
+
+        if (databaseActivity == null)
+        {
+            throw new ArgumentNullException(nameof(databaseActivity));
+        }
+
+        Logger = innerConnectionLifetime.Logger;
+        CommandTimeout = innerConnectionLifetime.CommandTimeout;
+        _innerConnectionLifetime = innerConnectionLifetime;
+        _openTelemetryOptions = openTelemetryOptions;
+        _databaseActivity = databaseActivity;
+    }
+
+    public ValueTask DisposeAsync()
+    {
+        return _innerConnectionLifetime.DisposeAsync();
+    }
+
+    public void Dispose()
+    {
+        _innerConnectionLifetime.Dispose();
+    }
+
+    public IMartenSessionLogger Logger { get; set; }
+    public int CommandTimeout { get; }
+    public int Execute(NpgsqlCommand cmd)
+    {
+        if (_openTelemetryOptions.TrackConnectionEvents)
+        {
+            _databaseActivity.AddEvent(new ActivityEvent(MartenCommandExecutionStarted));
+        }
+
+        try
+        {
+            return _innerConnectionLifetime.Execute(cmd);
+        }
+        catch (Exception e)
+        {
+            if (_openTelemetryOptions.TrackConnectionEvents)
+            {
+                _databaseActivity.AddEvent(new ActivityEvent(MartenCommandExecutionFailed, DateTimeOffset.UtcNow, new ActivityTagsCollection(new KeyValuePair<string, object>[] {new KeyValuePair<string, object>("ExceptionType", e.GetType()) })));
+            }
+
+            throw;
+        }
+    }
+
+    public Task<int> ExecuteAsync(NpgsqlCommand command, CancellationToken token = new CancellationToken())
+    {
+        if (_openTelemetryOptions.TrackConnectionEvents)
+        {
+            _databaseActivity.AddEvent(new ActivityEvent("Database command execution started"));
+        }
+
+        try
+        {
+            return _innerConnectionLifetime.ExecuteAsync(command, token);
+        }
+        catch (Exception e)
+        {
+            if (_openTelemetryOptions.TrackConnectionEvents)
+            {
+                _databaseActivity.AddEvent(new ActivityEvent(MartenCommandExecutionFailed, DateTimeOffset.UtcNow,
```

Comment:
Same here for `RecordException`.
If that can't be used, refactor here and above to a helper-method.